File: apps/app-frontend/src/helpers/content-favorites.ts
import { invoke } from '@tauri-apps/api/core'

export type FavoriteProvider = 'modrinth' | 'curseforge' | 'mcarchive'
export type FavoriteContentType = 'mod' | 'resourcepack' | 'datapack' | 'shader'

export interface ContentFavorite {
	provider: FavoriteProvider
	project_id: string
	content_type: FavoriteContentType
	saved_at: number
}

export interface ContentFavoriteInput {
	provider: FavoriteProvider
	project_id: string
	content_type: FavoriteContentType
}

export const FAVORITE_CONTENT_TYPES: FavoriteContentType[] = [
	'mod',
	'resourcepack',
	'datapack',
	'shader',
]

export function isFavoriteContentType(value: string): value is FavoriteContentType {
	return FAVORITE_CONTENT_TYPES.includes(value as FavoriteContentType)
}

export function contentFavoriteKey(provider: FavoriteProvider, projectId: string) {
	return `${provider}:${projectId}`
}

export async function listContentFavorites(): Promise<ContentFavorite[]> {
	return await invoke('plugin:content-favorites|content_favorites_list')
}

export async function addContentFavorite(favorite: ContentFavoriteInput): Promise<ContentFavorite> {
	return await invoke('plugin:content-favorites|content_favorites_add', { favorite })
}

export async function removeContentFavorite(
	provider: FavoriteProvider,
	projectId: string,
): Promise<void> {
	await invoke('plugin:content-favorites|content_favorites_remove', {
		provider,
		projectId,
	})
}

File: apps/app/src/api/content_favorites.rs
use crate::api::Result;
use theseus::content_favorites::{
    self, ContentFavorite, ContentFavoriteInput, ContentFavoriteProvider,
};

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("content-favorites")
        .invoke_handler(tauri::generate_handler![
            content_favorites_list,
            content_favorites_add,
            content_favorites_remove,
        ])
        .build()
}

#[tauri::command]
pub async fn content_favorites_list() -> Result<Vec<ContentFavorite>> {
    Ok(content_favorites::list().await?)
}

#[tauri::command]
pub async fn content_favorites_add(
    favorite: ContentFavoriteInput,
) -> Result<ContentFavorite> {
    Ok(content_favorites::add(favorite).await?)
}

#[tauri::command]
pub async fn content_favorites_remove(
    provider: ContentFavoriteProvider,
    project_id: String,
) -> Result<()> {
    content_favorites::remove(provider, &project_id).await?;
    Ok(())
}
